src/WalkingModule/src/Module.cpp

```diff
@@ -1751,3 +1790,60 @@ bool WalkingModule::stopWalking()
     m_robotState = WalkingFSM::Stopped;
     return true;
 }
+
+// This thread synchronizes the walking-controller with the navigation stack.
+// Writes on a port a boolean value when to replan the path
+void WalkingModule::computeNavigationTrigger()
+{
+    bool trigger = false;   //flag used to fire the wait for sending the navigation replanning trigger
+    yInfo() << "[WalkingModule::computeNavigationTrigger] Starting Thread";
+    yarp::os::NetworkClock myClock;
+    myClock.open("/clock", "/navigationTriggerClock");
+    bool enteredDoubleSupport = false, exitDoubleSupport = true;
+    while (m_runThreads)
+    {
+        // Block the thread until the robot is in the walking state
+        if (m_robotState != WalkingFSM::Walking)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000/m_navigationTriggerLoopRate));
+            continue;
+        }
+
+        //double support check
+        if (m_leftInContact.size()>0 && m_rightInContact.size()>0)  //external consistency check
+        {
+            if (m_leftInContact[0] && m_rightInContact[0])
+            {
+                if (exitDoubleSupport)
+                {
+                    enteredDoubleSupport = true;
+                    exitDoubleSupport = false;
+                }
+            }
+            else
+            {
+                if (! exitDoubleSupport)
+                {
+                    trigger = true; //in this way we have only one trigger each exit of double support
+                }
+                exitDoubleSupport = true;
+            }
+        }
+
+        //send the replanning trigger after a certain amount of seconds
+        if (trigger)
+        {
+            trigger = false;
+            //waiting -> could make it dependant by the current swing step duration
+            myClock.delay(m_navigationReplanningDelay);
+            yDebug() << "[WalkingModule::computeNavigationTrigger] Triggering navigation replanning";
+            auto& b = m_replanningTriggerPort.prepare();
+            b.clear();
+            b.add((yarp::os::Value)true);   //send the planning trigger
+            m_replanningTriggerPort.write();
+        }   
+        
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000/m_navigationTriggerLoopRate));
+    }
+    yInfo() << "[WalkingModule::computeNavigationTrigger] Terminating thread";
+}
```

Comment:
Please refrain from adding big chunks of code in the ``Module`` file as it is already cluttered and complicated to maintain. I would suggest moving this code to a dedicated class. You may take as an example the [``TrajectoryPlanner``](https://github.com/robotology/walking-controllers/tree/5d854b1c269e94d7e2eb9181d866e1a0f8416e62/src/TrajectoryPlanner) part. It lives in  a separate folder, in a separate class, and then it is exploited in the main ``Module`` class.